src/main/java/com/google/devtools/build/lib/remote/RemoteRepoContentsCacheImpl.java

```diff
@@ -0,0 +1,311 @@
+          }
+
+          @Override
+          public PathFragment localPathToExecPath(PathFragment localPath) {
+            throw new UnsupportedOperationException("Not used");
+          }
+
+          @Override
+          public SortedMap<PathFragment, ActionInput> getInputMapping(
+              SpawnRunner.SpawnExecutionContext context, boolean willAccessRepeatedly) {
+            throw new UnsupportedOperationException("Not used");
+          }
+        };
+    try {
+      // TODO: Consider uploading asynchronously.
+      UploadManifest.create(
+              cache.getRemoteCacheCapabilities(),
+              digestUtil,
+              remotePathResolver,
+              actionKey,
+              action,
+              COMMAND,
+              ImmutableList.of(fetchedRepoMarkerFile, fetchedRepoDir),
+              /* outErr= */ null,
+              /* exitCode= */ 0,
+              /* startTime= */ Instant.now(),
+              /* wallTimeMs= */ 0,
+              /* preserveExecutableBit= */ true)
+          .upload(context, cache, reporter);
+    } catch (ExecException | IOException e) {
+      reporter.handle(
+          Event.warn(
+              "Failed to upload repo contents to remote cache for repo %s: %s"
+                  .formatted(repoName, e.getMessage())));
+    }
+  }
+
+  @Override
+  public boolean lookupCache(
+      RepositoryName repoName,
+      Path repoDir,
+      String predeclaredInputHash,
+      ExtendedEventHandler reporter)
+      throws IOException, InterruptedException {
+    if (!(repoDir.getFileSystem() instanceof RemoteExternalOverlayFileSystem remoteFs)) {
+      return false;
+    }
+
+    var context = buildContext(repoName);
+    if (!context.getReadCachePolicy().allowRemoteCache()) {
+      return false;
+    }
+    var actionKey = new ActionKey(digestUtil.compute(buildAction(predeclaredInputHash)));
+    // The marker file is read right after and thus requested to be inlined.
+    var cachedActionResult =
+        cache.downloadActionResult(
+            context, actionKey, /* inlineOutErr= */ false, ImmutableSet.of(MARKER_FILE_PATH));
+    if (cachedActionResult == null) {
+      return false;
+    }
+
+    var actionResult = cachedActionResult.actionResult();
+    if (actionResult.getExitCode() != 0
+        || actionResult.getOutputFilesCount() != 1
+        || actionResult.getOutputDirectoriesCount() != 1) {
+      reporter.handle(
+          Event.warn(
+              String.format(
+                  "Unexpected action result for cached repo %s: exit code %d, %d output files, %d output directories",
+                  repoName,
+                  actionResult.getExitCode(),
+                  actionResult.getOutputFilesCount(),
+                  actionResult.getOutputDirectoriesCount())));
+      return false;
+    }
+
+    ListenableFuture<byte[]> markerFileContentFuture;
+    var markerFile = actionResult.getOutputFiles(0);
+    if (markerFile.getContents().isEmpty()) {
```

Comment:
how is this possible? is it because the "requested to be inlined" files are a suggestion only?